internal/builder/embedded/bkimage/authprovider.go

```diff
@@ -2,29 +2,65 @@ package bkimage
 
 import (
 	"context"
+	"crypto/ed25519"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"net/http"
+	"os"
+	"strconv"
+	"sync"
+	"time"
 
+	authutil "github.com/containerd/containerd/remotes/docker/auth"
+	remoteserrors "github.com/containerd/containerd/remotes/errors"
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/session/auth"
 	"github.com/pkg/errors"
+	"golang.org/x/crypto/nacl/sign"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
+// copied from https://github.com/moby/buildkit/blob/4e69662758446c7dc0e6de2bc1f7973d03bacbed/session/auth/authprovider/authprovider.go
+
+type tokenSeeds struct {
+	mu sync.Mutex
+	m  map[string]seed
+}
+
+type seed struct {
+	Seed []byte
+}
 type dynamicAuthProvider struct {
 	credFn CredentialsFn
+	seeds  *tokenSeeds
+}
+
+func NewDynamicAuthProvider(fn CredentialsFn) session.Attachable {
+	return &dynamicAuthProvider{
+		credFn: fn,
+		seeds:  &tokenSeeds{},
+	}
 }
 
 func (ap *dynamicAuthProvider) Register(server *grpc.Server) {
 	auth.RegisterAuthServer(server, ap)
 }
 
 func (ap *dynamicAuthProvider) Credentials(ctx context.Context, req *auth.CredentialsRequest) (*auth.CredentialsResponse, error) {
-	if req.Host == "registry-1.docker.io" {
-		req.Host = "https://index.docker.io/v1/"
+	return ap.credentials(req.Host)
```

Comment:
Do we really need the unexported `credentials` func here? It seems like (a) that func body is small and (b) this func only calls that one.

Reply:
I can remove it.
At the time the exported had a bit more code. 
There will be other changes to make it work since not all callers have a context.

Reply:
so Credentials, FetchToken and getAuthorityKey call `credentials`. The only thing they have in common is the host.